File: scratch/gen_hard_cases.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <fstream>
#include <cstdlib>
#include <ctime>

using namespace std;

int solve_query(const string& s, int L, int R) {
    string sub = s.substr(L, R - L + 1);
    int c = 0;
    for (char ch : sub) if (ch == '1') c++;
    
    vector<pair<char, int>> blocks;
    int i = 0;
    int n = sub.length();
    while (i < n) {
        char ch = sub[i];
        int start = i;
        while (i < n && sub[i] == ch) i++;
        blocks.push_back({ch, i - start});
    }
    
    int max_1s = c;
    for (int j = 1; j < (int)blocks.size() - 1; j++) {
        if (blocks[j].first == '1') {
            int L_i = blocks[j].second;
            int Z_L = blocks[j-1].second;
            int Z_R = blocks[j+1].second;
            
            int Z_max = 0;
            for (int k = 0; k < blocks.size(); k++) {
                if (blocks[k].first == '0') {
                    Z_max = max(Z_max, blocks[k].second);
                }
            }
            
            int new_Z = Z_L + L_i + Z_R;
            int gain = max(Z_max, new_Z) - L_i;
            max_1s = max(max_1s, c + gain);
        }
    }
    return max_1s;
}

vector<int> solve(const string& s, const vector<pair<int, int>>& queries) {
    vector<int> ans;
    for (auto& q : queries) {
        ans.push_back(solve_query(s, q.first, q.second));
    }
    return ans;
}

int main() {
    srand(time(NULL));
    
    string s = "1000100";
    vector<pair<int,int>> q = {{1,5},{0,6},{0,4}};
    vector<int> res = solve(s, q);
    for(int x : res) cout << x << " ";
    cout << endl;

    // Generate test cases
    ofstream out("scratch/expert_cases_new.json");
    out << "[\n";
    
    for (int t = 0; t < 15; t++) {
        int n, q_count;
        if (t < 4) { // medium
            n = 5000 + rand() % 1000;
            q_count = 5000 + rand() % 1000;
        } else { // hard
            n = 10000;
            q_count = 10000;
        }
        
        string s_gen = "";
        for (int i = 0; i < n; i++) {
            s_gen += (rand() % 2 == 0 ? '0' : '1');
        }
        
        vector<pair<int,int>> queries_gen;
        string queries_str = "[";
        for (int i = 0; i < q_count; i++) {
            int L = rand() % n;
            int R = rand() % n;
            if (L > R) swap(L, R);
            queries_gen.push_back({L, R});
            queries_str += "[" + to_string(L) + "," + to_string(R) + "]";
            if (i < q_count - 1) queries_str += ",";
        }
        queries_str += "]";
        
        vector<int> ans_gen = solve(s_gen, queries_gen);
        string ans_str = "[";
        for (int i = 0; i < q_count; i++) {
            ans_str += to_string(ans_gen[i]);
            if (i < q_count - 1) ans_str += ",";
        }
        ans_str += "]";
        
        out << "  {\n";
        out << "    \"input\": \"\\\"" << s_gen << "\\\"\\n" << queries_str << "\",\n";
        out << "    \"expectedOutput\": \"" << ans_str << "\"\n";
        out << "  }";
        if (t < 14) out << ",";
        out << "\n";
        
        cout << "Generated case " << t + 1 << endl;
    }
    
    out << "]\n";
    out.close();
    return 0;
}

File: scratch/gen_hard_cases.ts
import fs from 'fs';

function solveQuery(s: string, L: number, R: number): number {
    const sub = s.substring(L, R + 1);
    const n = sub.length;
    let c = 0;
    for (let i = 0; i < n; i++) {
        if (sub[i] === '1') c++;
    }

    const blocks: { char: string, len: number }[] = [];
    let i = 0;
    while (i < n) {
        const char = sub[i];
        const start = i;
        while (i < n && sub[i] === char) i++;
        blocks.push({ char, len: i - start });
    }

    let max_1s = c;
    for (let j = 1; j < blocks.length - 1; j++) {
        if (blocks[j].char === '1') {
            const L_i = blocks[j].len;
            const Z_L = blocks[j - 1].len;
            const Z_R = blocks[j + 1].len;

            let Z_max = 0;
            for (let k = 0; k < blocks.length; k++) {
                if (blocks[k].char === '0') {
                    if (blocks[k].len > Z_max) Z_max = blocks[k].len;
                }
            }

            const new_Z = Z_L + L_i + Z_R;
            const gain = Math.max(Z_max, new_Z) - L_i;
            if (c + gain > max_1s) {
                max_1s = c + gain;
            }
        }
    }
    return max_1s;
}

function solve(s: string, queries: number[][]): number[] {
    const ans = new Int32Array(queries.length);
    for (let i = 0; i < queries.length; i++) {
        ans[i] = solveQuery(s, queries[i][0], queries[i][1]);
    }
    return Array.from(ans);
}

function main() {
    const s = "1000100";
    const q = [[1, 5], [0, 6], [0, 4]];
    console.log(solve(s, q));

    const cases = [];
    for (let t = 0; t < 15; t++) {
        const isHard = t >= 4;
        const n = isHard ? 10000 : 5000 + Math.floor(Math.random() * 1000);
        const qCount = isHard ? 10000 : 5000 + Math.floor(Math.random() * 1000);

        let sGen = "";
        for (let i = 0; i < n; i++) {
            sGen += Math.random() < 0.5 ? '0' : '1';
        }

        const queriesGen = [];
        let queriesStr = "[";
        for (let i = 0; i < qCount; i++) {
            let L = Math.floor(Math.random() * n);
            let R = Math.floor(Math.random() * n);
            if (L > R) {
                const temp = L;
                L = R;
                R = temp;
            }
            queriesGen.push([L, R]);
            queriesStr += `[${L},${R}]`;
            if (i < qCount - 1) queriesStr += ",";
        }
        queriesStr += "]";

        const start = Date.now();
        const ansGen = solve(sGen, queriesGen);
        console.log(`Generated case ${t + 1} in ${Date.now() - start}ms`);

        let ansStr = "[";
        for (let i = 0; i < qCount; i++) {
            ansStr += ansGen[i];
            if (i < qCount - 1) ansStr += ",";
        }
        ansStr += "]";

        cases.push({
            input: `"${sGen}", ${queriesStr}`,
            expectedOutput: ansStr
        });
    }

    fs.writeFileSync("scratch/expert_cases_new.json", JSON.stringify(cases, null, 2));
    console.log("Done generating cases!");
}

main();
